### PR TITLE
Add linear interpolation to heat model evaluation

### LINKED ISSUE #74: Use linear interpolation for the heat model evaluation points
**Is your feature request related to a problem? Please describe.**
Right now, the heat model spatial evaluation points can only be grid points of the spatial discretization, which reduces flexibility. 

**Describe the solution you'd like**
Instead, linear interpolation should be used.

**Additional context**
The current implementation looks like this:

```python
# the indices of the wanted evaluation points
evaluation_indices = jnp.multiply(
    cls.evaluation_points,
    jnp.array(solution.shape),
).astype(int)

solution_at_evaluation_points = jnp.array(
    solution[
        evaluation_indices[:, 0],
        evaluation_indices[:, 1],
    ]
)
return solution_at_evaluation_points
```

### PR DESCRIPTION
**Pull request by castellinilinguini**
# Description

Add linear interpolation to heat model to increase flexibility with different evaluation points.

Fixes #74 .

## Type of change

Please tick at least one of the options. Delete the other options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please note/describe at least one tests that you ran to verify your changes.

- [x] Validated indices are correct

## Checklist For Contributor

- [x] My code follows the style guidelines of this project (I installed pre-commit)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote my changes in the changelog in the `[unreleased]` section
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

## Checklist For Maintainers

- [ ] Continuous Integration (CI) is successfully running
- [ ] Do we want to release/tag a new version? [✔️ / ❌]
  - [ ] If yes, add a release to the changelog and set the new version in pyproject.toml. After the merge, tag the release!

### CODE REVIEW

File: eulerpi/examples/heat/heat.py [hunk 1]
```diff
@@ -104,17 +105,31 @@ def forward(cls, param: np.ndarray) -> np.ndarray:
...
-
-        solution_at_evaluation_points = jnp.array(
-            solution[
-                evaluation_indices[:, 0],
-                evaluation_indices[:, 1],
-            ]
+        # linearly interpolate the solution at the evaluation points, use own interpolation function as jax doesn't support scipy.interpolate.interp2d and doesn't provide a 2d interpolation function
+        x = jnp.linspace(0, cls.plate_length[0], cls.num_grid_points)
+        y = jnp.linspace(0, cls.plate_length[1], cls.num_grid_points)
+        X, Y = jnp.meshgrid(x, y)
```
Comment (kaiserls) [inline eulerpi/examples/heat/heat.py:111]: X, Y are unused -> Remove them?

File: eulerpi/examples/heat/heat.py [hunk 2]
```diff
@@ -104,17 +105,31 @@ def forward(cls, param: np.ndarray) -> np.ndarray:
...
+            * (y[y_indices + 1] - cls.evaluation_points[:, 1])
+            + solution[x_indices + 1, y_indices]
+            * (cls.evaluation_points[:, 0] - x[x_indices])
+            * (y[y_indices + 1] - cls.evaluation_points[:, 1])
+            + solution[x_indices, y_indices + 1]
+            * (x[x_indices + 1] - cls.evaluation_points[:, 0])
+            * (cls.evaluation_points[:, 1] - y[y_indices])
+            + solution[x_indices + 1, y_indices + 1]
+            * (cls.evaluation_points[:, 0] - x[x_indices])
+            * (cls.evaluation_points[:, 1] - y[y_indices])
```
Comment (kaiserls) [inline eulerpi/examples/heat/heat.py:132]: I checked it against https://en.wikipedia.org/wiki/Bilinear_interpolation, looks good.

Depending on your time, you could refactor this into a function `interpolate_linear(x_eval, y_eval, x_grid, y_grid, func_grid)`, put it into the file `jax_extension.py` and test it using something like:
```python
n_eval = ...

f = lambda x,y:  2.0*x - y

x_grid = ...
y_grid = ...
X_grid, Y_grid = jnp.meshgrid(x_grid, y_grid)
f_grid = f(X_grid, Y_grid)

x_eval = np.random.random_sample((n_eval,))
y_eval = np.random.random_sample(n_eval,))
interp = interpolate_linear(...)

X_eval, Y_eval = jnp.meshgrid(x_eval, y_eval)
exact = f(X_eval, Y_eval)

assert jnp.allclose(interp, exact)
```

But that might already be over-engineering. I am totally fine with the current state of the pull request.

File: eulerpi/examples/heat/heat.py (hunk 1 above)
Comment (kaiserls) [inline eulerpi/examples/heat/heat.py:111]: X, Y are unused -> Remove them?

**Review by kaiserls — APPROVED** [review]
Really nice, just found some likely unused variables.

File: eulerpi/examples/heat/heat.py (hunk 1 above)
Comment (castellinilinguini) [reply eulerpi/examples/heat/heat.py:111]: Yep, just saw that a second too late, will do!

File: eulerpi/examples/heat/heat.py (hunk 2 above)
Comment (castellinilinguini) [reply eulerpi/examples/heat/heat.py:132]: Alright, I think I'll put that on the "future work" list :)

**Pull request merged**